tools/nimdigger.nim

```diff
@@ -0,0 +1,327 @@
+
+proc isGitNimTag(tag: string): bool =
+  if not tag.startsWith "v":
+    return false
+  let ver = tag[1..^1].split(".")
+  return ver.len == 3
+
+proc parseNimGitTag(tag: string): (int, int, int) =
+  doAssert tag.isGitNimTag, tag
+  let ver = tag[1..^1].split(".")
+  template impl(i) =
+    # improve pending https://github.com/nim-lang/Nim/pull/18038
+    result[i] = ver[i].parseInt
+  impl 0
+  impl 1
+  impl 2
+
+proc toNimCsourcesExe(binDir: string, name: string, rev: string): string =
+  let rev2 = rev.replace(".", "_")
+  result = binDir / fmt"nim_nimdigger_{name}_{rev2}{ExeExt2}"
+
+proc buildCsourcesRev(copt: CsourcesOpt) =
+  # sync with `_nimBuildCsourcesIfNeeded`
+  let csourcesExe = toNimCsourcesExe(copt.binDir, copt.name, copt.rev)
+  if csourcesExe.fileExists:
+    return
+  if verbose: dbg copt
+  if not copt.dir.dirExists: gitClone(copt.url, copt.dir)
+  if copt.fetch: gitFetch(copt.dir)
+  gitResetHard(copt.dir, copt.rev)
+  when defined(bsd):
+    let make = "gmake"
+  else:
+    let make = "make"
+  let oldNim = copt.binDir / "nim" & ExeExt2
+  removeFile(oldNim) # otherwise `make` may incorrectly decide there's notthing to build
+  let ncpu = countProcessors()
+  if copt.rev.isGitNimTag and copt.rev.parseNimGitTag < (0,15,2):
+    # avoids: make: *** No rule to make target `c_code/3_2/compiler_testability.o', needed by `../bin/nim'.  Stop.
+    discard runCmdOutput(fmt"sh build.sh {copt.csourcesBuildArgs}", copt.dir)
+  else:
+    discard runCmdOutput(fmt"{make} -j {ncpu + 2} -l {ncpu} {copt.csourcesBuildArgs}", copt.dir)
+  if isSimulate():
+    dbg csourcesExe
+  else:
+    copyFile(oldNim, csourcesExe)
+
+proc buildCsourcesAnyRevs(copt: CsourcesOpt) =
+  for rev in copt.revs:
+    copt.rev = rev
+    buildCsourcesRev(copt)
+
+proc toCsourcesRev(rev: string): string =
+  let ver = rev.parseNimGitTag
+  if ver >= (1, 0, 0): return csourcesRevs[^1]
+  for a in csourcesRevs[1 ..< ^1].reversed:
+    if ver >= a.parseNimGitTag: return a
+  return csourcesRevs[1] # because v0.9.4 seems broken
+
+proc getNimCsourcesAnyExe(state: DiggerState): CsourcesOpt =
+  let file = state.nimDir/"config/build_config.txt" # for newer nim versions, this file specifies correct csources_v1 to use
+  if file.fileExists:
+    let tab = file.readFile.parseKeyVal
+    result = state.coptv1
+    result.rev = tab["nim_csourcesHash"]
+  elif gitIsAncestorOf(state.nimDir, "a9b62de", state.rev): # commit that introduced csources_v1
+    result = state.coptv1
+    result.rev = csourcesV1Revs[0]
+  else:
+    let tag = gitLatestTag(state.nimDir)
+    result = state.coptv0
+    result.rev = tag.toCsourcesRev
+  result.nimCsourcesExe = toNimCsourcesExe(state.binDir, result.name, result.rev)
+
+proc main2(opt: DiggerOpt) =
+  let state = DiggerState(nimDir: opt.nimDir, rev: opt.rev)
+  if state.nimDir.len == 0:
+    let nimdiggerHome = getEnv(NimDiggerEnv, getHomeDir() / ".nimdigger")
```

Comment:
I thought about this but IMO the correct way to handle this would be via an addition to std/os, say ` getDataDir()` (we already have `getConfigDir()` for `XDG_CONFIG_HOME`, but no such equivalent for `XDG_DATA_DIR`). Once this is added, nimdigger can reuse it instead.

Note also that this follows the same pattern as `~/.nimble/nimblecache`

Reply:
now using instead `getEnv(NimDiggerEnv, getCacheDir("nimdigger"))` now that i've added os.getCacheDir